File: lib/types/Station.dart

import 'package:audio_service/audio_service.dart';
import 'package:flutter/cupertino.dart';
import 'package:get_it/get_it.dart';

import '../queries/getStations.graphql.dart';
import '../services/image_cache_service.dart';
import '../utils.dart';
import 'playlist_item.dart';

/// The playback kind of a station.
///
/// Mirrors the backend `station_type` wire value. Old cached payloads may omit
/// it entirely, in which case the station is treated as a [radio] stream.
enum StationMediaType {
  radio,
  tv,
  playlist;

  /// Parses a raw wire value case-insensitively, defaulting to [radio].
  ///
  /// Missing, null or unknown values fall back to [radio] so legacy caches and
  /// the bundled fallback asset keep working.
  static StationMediaType parse(String? value) {
    switch (value?.trim().toLowerCase()) {
      case 'tv':
        return StationMediaType.tv;
      case 'playlist':
        return StationMediaType.playlist;
      case 'radio':
      default:
        return StationMediaType.radio;
    }
  }
}

class Station {
  Query$GetStations$stations rawStationData;

  /// Aggregated review stats from REST API (reviews_stats field).
  /// These are separate from rawStationData.reviews because the REST endpoint
  /// returns pre-computed stats while the GraphQL reviews array may be empty.
  double _averageRating;
  int _numberOfReviews;

  Station({
    required this.rawStationData,
    double averageRating = 0,
    int numberOfReviews = 0,
  })  : _averageRating = averageRating,
        _numberOfReviews = numberOfReviews;

  get id => rawStationData.id;
  get slug => rawStationData.slug;
  get title => rawStationData.title;
  get order => rawStationData.order;
  get stationStreams => rawStationData.station_streams;
  get totalListeners => rawStationData.total_listeners;
  String? get thumbnailUrl => rawStationData.thumbnail_url;
  String get displayTitle => rawStationData.title;
  String get displaySubtitle => Utils.getCurrentPlayedSongTitle(rawStationData);
  String get artist => Utils.getCurrentPlayedSongTitle(rawStationData);
  bool get isUp => rawStationData.uptime?.is_up ?? false;
  int get songId => rawStationData.now_playing?.song?.id ?? -1;
  String get songTitle => rawStationData.now_playing?.song?.name ?? "";
  String get songArtist => rawStationData.now_playing?.song?.artist?.name ?? "";

  /// Playback kind of the station, defaulting to [StationMediaType.radio] when
  /// the backend omits `station_type` (old caches / fallback asset).
  StationMediaType get stationType =>
      StationMediaType.parse(rawStationData.station_type);

  /// Whether this station is a live TV channel.
  bool get isTv => stationType == StationMediaType.tv;

  /// Whether this station is an on-demand playlist.
  bool get isPlaylist => stationType == StationMediaType.playlist;

  /// Ordered playlist entries for this station.
  ///
  /// Returns an empty list when the station has no playlist (or the field is
  /// absent). Order is preserved exactly as returned by the API.
  List<PlaylistItem> get playlistItems {
    final items = rawStationData.playlist_items;
    if (items == null || items.isEmpty) return const [];
    return items.map(PlaylistItem.fromRaw).toList(growable: false);
  }

  ImageCacheService? get _imageCacheService {
    try {
      return GetIt.instance<ImageCacheService>();
    } catch (_) {
      return null;
    }
  }

  /// Cached path for the station's base thumbnail (stable URL).
  String? get cachedThumbnailPath {
    final url = rawStationData.thumbnail_url;
    if (url == null || url.isEmpty) return null;
    return _imageCacheService?.getCachedPath(url);
  }

  /// Cached path for the display thumbnail (song thumbnail if playing, else station thumbnail).
  String? get cachedArtPath {
    final displayUrl = Utils.getStationThumbnailUrl(rawStationData);
    if (displayUrl.isEmpty) return null;
    return _imageCacheService?.getCachedPath(displayUrl);
  }

  Uri get artUri {
    final cachedPath = cachedArtPath;
    if (cachedPath != null) {
      return Uri.file(cachedPath);
    }
    return Uri.parse(Utils.getStationThumbnailUrl(rawStationData));
  }

  Widget get thumbnail => displayThumbnail();

  Widget displayThumbnail({int? cacheWidth}) {
    final hasSongThumbnail = rawStationData.now_playing?.song?.thumbnail_url != null;
    return Utils.displayImage(
      Utils.getStationThumbnailUrl(rawStationData),
      fallbackImageUrl: rawStationData.thumbnail_url,
      cache: true,
      cachedFilePath: cachedArtPath,
      cacheWidth: cacheWidth,
      // Song thumbnails expire after 10 minutes; station logos cached permanently
      cacheMaxAge: hasSongThumbnail ? const Duration(minutes: 10) : null,
    );
  }

  /// Average star rating from reviews_stats (REST API pre-computed).
  /// Falls back to computing from individual reviews if stats aren't available.
  double get averageRating {
    if (_averageRating > 0) return _averageRating;
    final reviews = rawStationData.reviews;
    if (reviews.isEmpty) return 0;
    final sum = reviews.fold<int>(0, (s, r) => s + r.stars);
    return sum / reviews.length;
  }

  /// Number of reviews from reviews_stats (REST API pre-computed).
  int get reviewCount {
    if (_numberOfReviews > 0) return _numberOfReviews;
    return rawStationData.reviews.length;
  }

  MediaItem get mediaItem {
    return MediaItem(
      id: Utils.getStationStreamUrls(rawStationData).firstOrNull ?? "",
      title: rawStationData.title,
      displayTitle: displayTitle,
      displaySubtitle: displaySubtitle,
      artist: artist,
      duration: null,
      artUri: artUri,
      isLive: true,
      extras: {
        "station_id": rawStationData.id,
        "station_slug": rawStationData.slug,
        "station_title": rawStationData.title,
        "station_type": stationType.name,
        "song_id": songId,
        "song_title": songTitle,
        "song_artist": songArtist,
        "total_listeners": rawStationData.total_listeners,
        "station_is_up": isUp,
        "station_thumbnail_url": rawStationData.thumbnail_url,
        "station_streams": Utils.getStationStreamObjects(rawStationData),
      },
    );
  }

}

File: apple_tv/RadioCrestinTV/Sources/Models/Station.swift
import Foundation

/// Pared-down model — only the fields the tvOS UI actually consumes.
/// Mirrors the shape returned by `GET /api/v1/stations` so the same
/// REST responses parse without server-side changes. Extra keys present
/// on the wire (description, posts, facebook_page_id, …) are ignored.
struct Station: Codable, Identifiable, Hashable {
    let id: Int
    let slug: String
    let title: String
    let order: Int
    let thumbnailUrl: String?
    var totalListeners: Int?

    let stationStreams: [StationStream]
    var uptime: Uptime?
    var nowPlaying: NowPlaying?
    let reviewsStats: ReviewsStats?

    /// Backend-provided station kind. Optional so older API responses that
    /// predate the field still decode — a missing/null value means radio.
    let stationType: String?

    /// Ordered, enabled playlist entries for `station_type == "playlist"`.
    /// Optional and defensively decoded — radio/TV stations omit it.
    let playlistItems: [PlaylistItem]?

    /// DVR (timeshift) variant of the live HLS stream — a 1-hour sliding
    /// window that carries the same `EXT-X-PROGRAM-DATE-TIME` + `DATERANGE`
    /// tags as the short-window live index. Nullable and often absent
    /// (absent entirely in production today).
    ///
    /// Decoded only to future-proof the model against the backend contract;
    /// **DVR/timeshift playback is intentionally NOT implemented on tvOS
    /// yet**, so nothing currently reads this field. Kept so a later feature
    /// can wire up live rewind without another model migration.
    let hlsDvrStreamUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, slug, title, order
        case thumbnailUrl = "thumbnail_url"
        case totalListeners = "total_listeners"
        case stationStreams = "station_streams"
        case uptime
        case nowPlaying = "now_playing"
        case reviewsStats = "reviews_stats"
        case stationType = "station_type"
        case playlistItems = "playlist_items"
        case hlsDvrStreamUrl = "hls_dvr_stream_url"
    }

    /// Whether the station is currently reachable. A **missing** `uptime`
    /// record — the backend omits it when `check_uptime == false` (private
    /// dev/test stations, and some real TV/playlist stations) — means
    /// *unknown*, not down: it reads as `true` so the station stays playable
    /// and shows no warning. Only an explicit `is_up == false` marks a
    /// station as unavailable.
    var isUp: Bool { uptime?.isUp ?? true }

    /// How this station should be played. Parsed case-insensitively;
    /// anything unrecognized (or absent) falls back to `.radio`.
    var kind: StationKind { StationKind(rawWireValue: stationType) }

    /// Playlist entries the tvOS player can actually render, in **wire
    /// order** — the backend serves them newest-first, and the first
    /// array element must play first / show at the top, so we filter
    /// only and never re-sort. Allowlist-based: only `audio` and `video`
    /// survive — YouTube entries (single clips or whole playlists) and
    /// any unknown future type strings cannot be embedded on tvOS and
    /// would only hand AVPlayer an unplayable URL.
    var playableItems: [PlaylistItem] {
        (playlistItems ?? []).filter(\.isPlayable)
    }

    /// True when this is a playlist station whose entries exist but none
    /// is playable on tvOS (YouTube-only lists, unknown types). The UI
    /// shows a friendly explanation instead of a dead player.
    var hasOnlyUnplayableItems: Bool {
        guard kind == .playlist else { return false }
        let items = playlistItems ?? []
        return !items.isEmpty && !items.contains(where: \.isPlayable)
    }

    /// Value reported to analytics for the station kind ("radio"/"tv"/
    /// "playlist"). Never nil so dashboards can group cleanly.
    var kindAnalyticsValue: String { kind.rawValue }

    /// Streams in API order (the audio handler tries them sequentially on
    /// playback failure — the same fallback contract Android implements).
    var orderedStreams: [StationStream] {
        stationStreams.sorted { $0.order < $1.order }
    }

    /// True when the primary (lowest-order) stream is HLS. Used by the
    /// metadata sync to decide whether an offset timestamp is needed.
    var primaryStreamIsHls: Bool {
        orderedStreams.first?.type == "HLS"
    }

    var songTitle: String { nowPlaying?.song?.name ?? "" }
    var songArtist: String { nowPlaying?.song?.artist?.name ?? "" }
    var songThumbnailUrl: String? {
        nowPlaying?.song?.thumbnailUrl ?? nowPlaying?.song?.artist?.thumbnailUrl
    }

    /// Secondary line shown under the station title in the grid. Radio
    /// stations surface the current song and fall back to an empty string
    /// when nothing is playing (the grid then shows only the title).
    /// Playlist and TV stations never carry a `now_playing` song, so
    /// rather than render an empty line they get a typed descriptor: a
    /// playlist summary (playable-item count, or a plain label when there's
    /// nothing to count) or a live-TV label.
    var cardSubtitle: String {
        switch kind {
        case .radio:
            return songTitle
        case .tv:
            return "Transmisiune live"
        case .playlist:
            let count = playableItems.count
            guard count > 0 else { return "Listă de redare" }
            return count == 1
                ? "Listă de redare · o piesă"
                : "Listă de redare · \(count) piese"
        }
    }

    /// URL list to try in order: song / artist art first (so the user
    /// sees the actual album cover when available), then station logo.
    /// Mirrors Station.displayThumbnail in the Flutter app.
    var displayThumbnailURLs: [URL] {
        [songThumbnailUrl, thumbnailUrl]
            .compactMap { $0 }
            .compactMap { URL(string: $0) }
    }

    /// Server-computed average rating (0..5). 0 when no reviews.
    var averageRating: Double {
        reviewsStats?.averageRating ?? 0
    }

    var reviewCount: Int {
        reviewsStats?.numberOfReviews ?? 0
    }

    /// Review-based score used by the recommended sort: average × count.
    /// Stations with many high reviews float to the top.
    var reviewScore: Double {
        averageRating * Double(reviewCount)
    }

    /// Returns a copy with `now_playing`, `uptime`, and listener count
    /// replaced from the `/stations-metadata` payload. Mirrors the Flutter
    /// `_mergeStationWithMetadata` logic — `liveListeners` (when supplied)
    /// always wins over the offset metadata's listener count so the
    /// audience number reflects what's actually listening *now* rather
    /// than what was listening at the HLS playback timestamp.
    func merging(_ metadata: StationMetadata, liveListeners: Int?) -> Station {
        var copy = self
        if let m = metadata.uptime {
            copy.uptime = m
        }
        if let np = metadata.now_playing {
            copy.nowPlaying = NowPlaying(
                id: nowPlaying?.id ?? 0,
                timestamp: np.timestamp,
                song: np.song
            )
            copy.totalListeners = liveListeners ?? np.listeners ?? totalListeners
        }
        return copy
    }
}

/// The three playback experiences a station can offer.
enum StationKind: String, Hashable {
    case radio
    case tv
    case playlist

    /// Maps the raw `station_type` wire string (any casing) to a kind.
    /// Unknown / missing / null all collapse to `.radio`.
    init(rawWireValue: String?) {
        switch rawWireValue?.lowercased() {
        case "tv": self = .tv
        case "playlist": self = .playlist
        default: self = .radio
        }
    }
}

/// One entry in a playlist station. The wire delivers entries
/// newest-first (descending `playlist_item_order`) and enabled-only;
/// apps must preserve that order exactly, so nothing re-sorts by
/// `order`. `type` is `audio`, `video`, `youtube`, or
/// `youtube_playlist` (the YouTube kinds — and any future type
/// strings — are unplayable on tvOS).
struct PlaylistItem: Codable, Hashable, Identifiable {
    let id: Int
    let order: Int
    let type: String
    let url: String
    let title: String
    let thumbnailUrl: String?
    let durationSeconds: Int?

    enum CodingKeys: String, CodingKey {
        case id, order, type, url, title
        case thumbnailUrl = "thumbnail_url"
        case durationSeconds = "duration_seconds"
    }

    var isVideo: Bool { type.lowercased() == "video" }
    var isAudio: Bool { type.lowercased() == "audio" }

    /// True when tvOS can render this item natively. Allowlist — only
    /// `audio` and `video` qualify, so `youtube`, `youtube_playlist`,
    /// and any unknown future type degrade to "not playable" instead of
    /// handing AVPlayer a URL it cannot open.
    var isPlayable: Bool { isAudio || isVideo }
}

struct ReviewsStats: Codable, Hashable {
    let averageRating: Double
    let numberOfReviews: Int

    enum CodingKeys: String, CodingKey {
        case averageRating = "average_rating"
        case numberOfReviews = "number_of_reviews"
    }
}

struct StationStream: Codable, Hashable {
    let order: Int
    let type: String
    let streamUrl: String

    enum CodingKeys: String, CodingKey {
        case order, type
        case streamUrl = "stream_url"
    }

    var isHls: Bool { type == "HLS" }
}

struct Uptime: Codable, Hashable {
    let isUp: Bool
    let timestamp: String?

    enum CodingKeys: String, CodingKey {
        case isUp = "is_up"
        case timestamp
    }
}

struct NowPlaying: Codable, Hashable {
    let id: Int
    let timestamp: String?
    let song: Song?
}

struct Song: Codable, Hashable {
    let id: Int
    let name: String
    let thumbnailUrl: String?
    let artist: Artist?

    enum CodingKeys: String, CodingKey {
        case id, name, artist
        case thumbnailUrl = "thumbnail_url"
    }
}

struct Artist: Codable, Hashable {
    let id: Int
    let name: String
    let thumbnailUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, name
        case thumbnailUrl = "thumbnail_url"
    }
}
